### ISSUE TITLE
Loss of confidence in Rasa > 1.6.0 nlu (compared to 1.4.6)

### ISSUE THREAD

**Issue by samscudder** [RasaHQ/rasa#5004]
**Rasa version**: 1.4.6 / 1.6.0

**Rasa SDK version** (if used & relevant):

**Rasa X version** (if used & relevant):

**Python version**: 3.6.9 (anaconda)

**Operating system** (windows, osx, ...): RHEL 7 (Centos)

**Issue**: After upgrading from Rasa 1.4.6 to 1.6.0, the latest version trains a lot faster (05:23.830 against 12:48.062 for 1.4.6). The speed gain is in the EmbeddingIntentClassifier, but confidence has fallen dramatically.

Our chatbot answer certain topics differently depending on the user profile, which is stored in a categorial slot. For testing purposes we have an intent that allows you to select the profile (using buttons). We also have an intent to show what profile is currently selected. For the comparison, I set up two fresh machines, with exactly the same Rasa configuration and training data. Here's a comparison of the accuracy rounded to 3 decimal places:

**With rasa shell --debug:**

"Select profile" intent - version 1.4.6 -> 0.980 - version 1.6.0 -> 0.659
"Who am I?" intent - version 1.4.6 -> 0.824 - version 1.6.0 -> 0.589

We use a fallback threshold of 0.7, so, with these results, no matter what we enter in the chatbot, we get the fallback action.

Our dataset contains 218 distinct intents.

**Error (including full traceback)**:
```
DNA
```

**Command or request that led to error**:
```
rasa shell --endpoints endpoints.yml --debug
```

**Content of configuration file (config.yml)** (if relevant):
```yml
# Configuration for Rasa NLU.
# https://rasa.com/docs/rasa/nlu/components/
language: pt

pipeline:
- name: "SpacyNLP"
- name: "SpacyTokenizer"
- name: "RegexFeaturizer"
- name: "CRFEntityExtractor"
- name: "EntitySynonymMapper"
- name: "CountVectorsFeaturizer"
  strip_accents: "unicode"
  analyzer: "char_wb"
  min_ngram: 1
  max_ngram: 4
- name: "CustomLemmatization"
  use_cls_token: false
- name: "EmbeddingIntentClassifier"


# Configuration for Rasa Core.
# https://rasa.com/docs/rasa/core/policies/
policies:
  - name: FormPolicy
  - name: AugmentedMemoizationPolicy
    max_history: 1
  - name: KerasPolicy
    validation_split: 0.2
    batch_size: 100
    epochs: 300
  - name: MappingPolicy
  - name: "FallbackPolicy"
    nlu_threshold: 0.7
    core_threshold: 0.7
    fallback_action_name: "utter_default"
```

**Content of domain file (domain.yml)** (if relevant):
```yml

```

**Comment by sara-tagger** [comment]
Thanks for raising this issue, @erohmensing will get back to you about it soon✨

###### Please also check out the [docs](https://rasa.com/docs/) and the [forum](https://forum.rasa.com/) in case your issue was raised there too 🤗

**Comment by JulianGerhard21** [comment]
Hi @samscudder,

I smell that this has something to do with a [change in version 1.5.0](https://rasa.com/docs/rasa/changelog/#id52).

If you want to test if I am correct, simply rewrite line `191` in "`nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py`" of rasa core to:

```
return (
            MESSAGE_ATTRIBUTES if analyzer in ["word", "char_wb"] else DENSE_FEATURIZABLE_ATTRIBUTES
        )
```

and then evaluate again - the results might be better then. I observed something similar but unfortunately didn't have the time to get a solution. Since Rasa did this on purpose, the workaround should be seen as only temporary.

**Comment by erohmensing** [comment]
In order to also check if it was a change in 1.5.0, could you run 1.5.0 to see what the confidence level of that is? Ideally also set the same `random_seed` in your configuration when testing confidences across the different versions.

**Comment by samscudder** [comment]
I'm setting up a machine with Rasa 1.5.0 right now

**Comment by samscudder** [comment]
With rasa 1.5.0 - 04m32s to train. Confidence of the intents I was comparing is 0.646 and 0.644, so I think it's safe to say it was a change in 1.5.0.

@erohmensing Good call on the random_seed, and we will be using that in the future, but typically the difference we noticed is something like 1-2%.

**Comment by samscudder** [comment]
@JulianGerhard21 Hit the nail on the head. With the patch, confidence went up to 0.94 / 0.97 again.

**Comment by khaleelp** [comment]
@sara-tagger  Will be any fix for this from Rasa or should we go ahead with fix which @JulianGerhard21 suggested?

**Comment by samscudder** [comment]
Patch suggested by @JulianGerhard is temporary. This bug only affects people using CountVectorFeaturizer with char_wb analysis.

**Comment by erohmensing** [comment]
@samscudder thanks for pointing this out. The idea is that the names of intents are not necessarily composed of meaningful words, so character level featurization might not make sense. We're considering making that featurization optional, though. Are you able to share your list of intents so I can look at the format and how it might be affecting the confidence?

**Comment by samscudder** [comment]
@erohmensing we are recognizing in portuguese (brazilian). We tried the other analysis options (word and char), and the confidence was lower than char_wb (in version 1.4.6). If you want, I can email you a spreadsheet with a list of phrases and the expected intent, and results using models trained with the different analysis options.

**Comment by erohmensing** [comment]
for that line of code the phrases are not important, just the intent names themselves. do you have different words of your intents separated, e.g. `in_snake_case`? or do you have them `concatenatedlikethis`?

I think the issue is that you actually want one more CountVectorsFeaturizer in your pipeline. Right now you're only using char_wb, and not word at all -- then you're losing not only on intent features, but also on word embeddings in your phrases. Our [default pipeline](https://rasa.com/docs/rasa/nlu/choosing-a-pipeline/#section-supervised-embeddings-pipeline) uses both, i.e. one that does word embeddings and the other that does character embeddings. I think this will help a lot. 

In other words, change to
```
pipeline:
- name: "SpacyNLP"
- name: "SpacyTokenizer"
- name: "RegexFeaturizer"
- name: "CRFEntityExtractor"
- name: "EntitySynonymMapper"
- name: "CountVectorsFeaturizer"
- name: "CountVectorsFeaturizer"
  strip_accents: "unicode"
  analyzer: "char_wb"
  min_ngram: 1
  max_ngram: 4
- name: "CustomLemmatization"
  use_cls_token: false
- name: "EmbeddingIntentClassifier"
```

**Comment by samscudder** [comment]
All our intent names are in snake_case.
 
I ran our current test suite (104 phrases, of which 25 are expected to fail due to mispelling or slang) with three different config files on Rasa 1.4.6:

- With one featurizer (previous config), we had a score of 73 correct classifications.
- With two featurizers (using your suggested config), the score increased to 77 correct.
- With two featurizers (and "strip accents" on both), the score hit 84, which is higher than the expected . 

Our test suite won't run on 1.6.0 due to a tracker bug I picked up yesterday. I'm waiting for it to be fixed to generate results for 1.6.0 for comparison.

**Comment by erohmensing** [comment]
Ah sure, sorry, forgot about adding stripping of constants to the other one, that makes plenty of sense.

> Our test suite won't run on 1.6.0 due to a tracker bug I picked up yesterday. I'm waiting for it to be fixed to generate results for 1.6.0 for comparison.

Gotcha. Really awesome that you caught that and were able to describe so well the behavior that was happening -- theres already a PR for a fix :)

**Comment by samscudder** [comment]
I just ran a reduced test suite against 1.4.6 and 1.6.2, 102 test phrases. I used the same random seed. rasa database was dropped between tests.

Our config file:
Keras policy is batch=128 epochs=600
EmbeddingIntentClassifier with 200 epochs

**Rasa 1.4.6**
Training time:  - Between 2 and 3 hours*
Model file size: 27.5MB
Correct intents (threshold 0.6): 58/102
Average accuracy of correct intents: 0.87

**Rasa 1.6.2 (stock)**
Training time: 7m40.77
Model file size: 51.8MB
Correct intents (threshold 0.6): 13/102
Average accuracy of correct intents: 0.71

**Rasa 1.6.2 (with @JulianGerhard21 patch)**
Training time: 8m09.75
Model file size: 53.4MB
Correct intents (threshold 0.6): 16/102
Average accuracy of correct intents: 0.66

* This was a rough estimate as our server was predicting 3 hours for the EmbeddingIntentClassifier, so I trained on a better machine which took around 20 minutes.

**Comment by JulianGerhard21** [comment]
Hi @samscudder 

I am currently investigating this with Rasa 1.7.0. Ill come back with some stats asap.

Regards

**Comment by samscudder** [comment]
Here's the results for Rasa 1.7.0 with the same training parameters and test cases as above:

**Stock rasa 1.7.0**
Training time: 7m40.55
Model file size: 51.9MB
Correct intents (threshold 0.6): 32/102
Average accuracy of correct intents: 0,69

**Rasa 1.7.0 with count_vector_featurizer patch**
Training time: 8:12.96
Model file size: 53.5MB
Correct intents (threshold 0.6): 29/102
Average accuracy of correct intents: 0,67

**Comment by JulianGerhard21** [comment]
Hi @samscudder ,

I am wondering about those results. With version 1.7, the patch is no longer necessary due to [this issue](https://github.com/rasahq/rasa/issues/4978). 

I tested a comparison between different sized datasets and configs over version 1.6.1 with patch and 1.7.0 without and the results were mostly better than expected, generally way over 90% F1 score since my sets are almost always imbalanced.

Can you tell me a bit more about your config and dataset? I am beginning to think that another approach might suit your problem better...

Kind regards
Julian

**Comment by erohmensing** [comment]
@samscudder i'm trying to understand how you're presenting your results. 
`Correct intents (threshold 0.6): 32/102`  -- does this mean you are only including those intents that were classified correctly and above the threshold here? 

Going from 1.4.6 to 1.6.2 stock [here](https://github.com/RasaHQ/rasa/issues/5004#issuecomment-579948544), is the decrease in correct intents due to new misclassifications or lower confidence?

**Comment by samscudder** [comment]
@erohmensing I just trained in 1.4.6 and 1.7.0. The results are slightly different from above as I'm using only 220 epochs for the EmbeddingIntentClassifier, and 200 epochs for Keras. Here is a detailed report.

Using this `config.yaml`:
```
# Configuration for Rasa NLU.
# https://rasa.com/docs/rasa/nlu/components/
language: pt

pipeline:
- name: "SpacyNLP"
- name: "SpacyTokenizer"
- name: "RegexFeaturizer"
- name: "CRFEntityExtractor"
- name: "CountVectorsFeaturizer"
  strip_accents: "unicode"
- name: "CountVectorsFeaturizer"
  strip_accents: "unicode"
  analyzer: "char_wb"
  min_ngram: 1
  max_ngram: 4
- name: "components.lemmatization.lemma.CustomLemmatization"
  use_cls_token: false
- name: "EmbeddingIntentClassifier"
  epochs: 220
  random_seed: 2614

# Configuration for Rasa Core.
# https://rasa.com/docs/rasa/core/policies/
policies:
  - name: FormPolicy
  - name: AugmentedMemoizationPolicy
    max_history: 1
  - name: KerasPolicy
    random_seed: 2614
    batch_size: 64
    epochs: 200
  - name: MappingPolicy
  - name: "FallbackPolicy"
    nlu_threshold: 0.6
    core_threshold: 0.6
    fallback_action_name: action_padrao
```
The dataset has 273 intents, 356 actions, 551 stories, and 2694 phrases in the nlu.

Training time in 1.4.6 (8 i7 cores + rtx1050):  5m17.948s

I'll post an abridged training output in the next post.

**The results for 1.4.6 were:**

102 phrases, where 25 are purposely expected to fail.
9 phrases classified wrong, 3 of these with a confidence > 0.6
75 phrases were correctly classified, with a confidence > 0.6
Average overall confidence was 0.7914
Average confidence of the 78 phrases above 0.6 was 0.9254

In Rasa 1.7, on the same machine, training time was 3m14.874s

**The results for 1.7.0 were:**

The same 102 phrases, where 25 are purposely expected to fail.
25 phrases classified wrong, none with confidence > 0.6
61 phrases were correctly classified, with a confidence > 0.6
Average overall confidence was 0.7024
Average confidence of the 61 phrases above 0.6 was 0,84

To get the same results, I have to reduce the threshold to 0,35. But then I have 16 wrongly classified phrases above threshold.

I'm sending 102 test phrases to the API, and getting `latest-message->intent_ranking->0->Name` and `latest-message->intent_ranking->0->confidence` from the tracker after each phrase. I have a spreadsheet of all expected results, and I compare which are correctly classified, and which have a confidence > threshold.

Environment was creates with this:

```
conda create -n rasa17 python=3.6.9 anaconda
conda activate rasa17
pip install rasa[spacy]
conda install cudnn
pip uninstall tensorflow-cpu==1.15.0
pip install tensorflow==1.15.0
python -m spacy download pt_core_news_sm
python -m spacy link pt_core_news_sm pt --force
python
>>> import spacy
>>> spacy.load('pt')
>>> exit()
```

I also had to change from `from rasa.nlu.tokenizers import Token, Tokenizer` to `from rasa.nlu.tokenizers.tokenizer import Token, Tokenizer` in our lemma script.

**Comment by samscudder** [comment]
Abridged training output for above:

**Rasa 1.4.6**
```
Training Core model...
2020-02-06 11:42:26 INFO     numexpr.utils  - NumExpr defaulting to 8 threads.
2020-02-06 11:42:28 INFO     absl  - Entry Point [tensor2tensor.envs.tic_tac_toe_env:TicTacToeEnv] registered with id [T2TEnv-TicTacToeEnv-v0]
Processed Story Blocks: 100%|█████████████████████████| 549/549 [00:00<00:00, 3098.28it/s, # trackers=1]
Processed Story Blocks: 100%|██████████████████████████| 549/549 [00:06<00:00, 90.43it/s, # trackers=50]
Processed Story Blocks: 100%|██████████████████████████| 549/549 [00:07<00:00, 76.33it/s, # trackers=50]
Processed Story Blocks: 100%|██████████████████████████| 549/549 [00:06<00:00, 81.87it/s, # trackers=46]
Processed trackers: 100%|███████████████████████████| 544/544 [00:00<00:00, 3162.95it/s, # actions=1755]
Processed trackers: 100%|███████████████████████████| 544/544 [00:00<00:00, 3201.02it/s, # actions=1706]
Processed actions: 1706it [00:00, 19612.93it/s, # examples=1698]
Processed trackers: 100%|█████████████████████████| 1044/1044 [00:00<00:00, 1799.68it/s, # actions=3584]
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #
=================================================================
masking (Masking)            (None, 5, 656)            0
_________________________________________________________________
lstm (LSTM)                  (None, 32)                88192
_________________________________________________________________
dense (Dense)                (None, 368)               12144
_________________________________________________________________
activation (Activation)      (None, 368)               0
=================================================================
Total params: 100,336
Trainable params: 100,336
Non-trainable params: 0
_________________________________________________________________
2020-02-06 11:42:51 INFO     rasa.core.policies.keras_policy  - Fitting model with 3584 total samples and a validation split of 0.1
Train on 3584 samples
Epoch 1/200
3584/3584 [==============================] - 1s 239us/sample - loss: 5.2628 - acc: 0.3683
.
.
.
Epoch 200/200
3584/3584 [==============================] - 0s 99us/sample - loss: 0.3145 - acc: 0.9381
2020-02-06 11:44:04 INFO     rasa.core.policies.keras_policy  - Done fitting keras policy model
2020-02-06 11:44:06 INFO     rasa.core.agent  - Persisted model to '/tmp/tmp36v9kcw4/core'
Core model training completed.
Training NLU model...
2020-02-06 11:44:06 INFO     rasa.nlu.utils.spacy_utils  - Trying to load spacy model with name 'pt'
2020-02-06 11:44:07 INFO     rasa.nlu.components  - Added 'SpacyNLP' to component cache. Key 'SpacyNLP-pt'.
2020-02-06 11:44:07 INFO     rasa.nlu.training_data.training_data  - Training data stats:
        - intent examples: 2693 (226 distinct intents)
        - Found intents: ...
        - Number of response examples: 0 (0 distinct response)
        - entity examples: 39 (1 distinct entities)
        - found entities: ...

2020-02-06 11:44:07 INFO     rasa.nlu.model  - Starting to train component SpacyNLP
2020-02-06 11:44:10 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 11:44:10 INFO     rasa.nlu.model  - Starting to train component SpacyTokenizer
2020-02-06 11:44:10 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 11:44:10 INFO     rasa.nlu.model  - Starting to train component RegexFeaturizer
2020-02-06 11:44:10 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 11:44:10 INFO     rasa.nlu.model  - Starting to train component CRFEntityExtractor
2020-02-06 11:44:11 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 11:44:11 INFO     rasa.nlu.model  - Starting to train component CountVectorsFeaturizer
2020-02-06 11:44:11 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 11:44:11 INFO     rasa.nlu.model  - Starting to train component CountVectorsFeaturizer
2020-02-06 11:44:12 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 11:44:12 INFO     rasa.nlu.model  - Starting to train component tokenizer_spacy_lemma
2020-02-06 11:44:12 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 11:44:12 INFO     rasa.nlu.model  - Starting to train component EmbeddingIntentClassifier
Epochs: 100%|██████████████████████████████████| 220/220 [03:22<00:00,  1.08it/s, loss=0.960, acc=0.998]
2020-02-06 11:47:35 INFO     rasa.utils.train_utils  - Finished training embedding policy, train loss=0.960, train accuracy=0.998
2020-02-06 11:47:35 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 11:47:36 INFO     rasa.nlu.model  - Successfully saved model into '/tmp/tmp36v9kcw4/nlu'
NLU model training completed.
```

**Rasa 1.7.0**
```
Training Core model...
2020-02-06 12:31:30 INFO     numexpr.utils  - NumExpr defaulting to 8 threads.
2020-02-06 12:31:32 INFO     absl  - Entry Point [tensor2tensor.envs.tic_tac_toe_env:TicTacToeEnv] registered with id [T2TEnv-TicTacToeEnv-v0]
/home/viceri/anaconda3/envs/rasa17/lib/python3.6/site-packages/rasa/core/domain.py:130: FutureWarning: Your domain file contains the key: 'templates'. This has been deprecated and renamed to 'responses'. The 'templates' key will no longer work in future versions of Rasa. Please replace 'templates' with 'responses'
  return cls.from_dict(data)
/home/viceri/anaconda3/envs/rasa17/lib/python3.6/site-packages/rasa/core/domain.py:148: FutureWarning: No tracker session configuration was found in the loaded domain. Domains without a session config will automatically receive a session expiration time of 60 minutes in Rasa version 2.0 if not configured otherwise.
  session_config = cls._get_session_config(data.get(SESSION_CONFIG_KEY, {}))
Processed Story Blocks: 100%|█████████████████████████| 549/549 [00:00<00:00, 2910.83it/s, # trackers=1]
Processed Story Blocks: 100%|██████████████████████████| 549/549 [00:06<00:00, 87.46it/s, # trackers=50]
Processed Story Blocks: 100%|██████████████████████████| 549/549 [00:07<00:00, 77.75it/s, # trackers=50]
Processed Story Blocks: 100%|██████████████████████████| 549/549 [00:07<00:00, 74.61it/s, # trackers=48]
Processed trackers: 100%|███████████████████████████| 544/544 [00:00<00:00, 2529.18it/s, # actions=1755]
Processed trackers: 100%|███████████████████████████| 544/544 [00:00<00:00, 2561.15it/s, # actions=1706]
Processed actions: 1706it [00:00, 19692.33it/s, # examples=1698]
Processed trackers: 100%|█████████████████████████| 1044/1044 [00:00<00:00, 1455.60it/s, # actions=3615]
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #
=================================================================
masking (Masking)            (None, 5, 659)            0
_________________________________________________________________
lstm (LSTM)                  (None, 32)                88576
_________________________________________________________________
dense (Dense)                (None, 371)               12243
_________________________________________________________________
activation (Activation)      (None, 371)               0
=================================================================
Total params: 100,819
Trainable params: 100,819
Non-trainable params: 0
_________________________________________________________________
2020-02-06 12:31:56 INFO     rasa.core.policies.keras_policy  - Fitting model with 3615 total samples and a validation split of 0.1
Train on 3615 samples
Epoch 1/200
3615/3615 [==============================] - 1s 238us/sample - loss: 5.2815 - acc: 0.3685
.
.
.
Epoch 200/200
3615/3615 [==============================] - 0s 99us/sample - loss: 0.3115 - acc: 0.9364
2020-02-06 12:33:09 INFO     rasa.core.policies.keras_policy  - Done fitting keras policy model
2020-02-06 12:33:11 INFO     rasa.core.agent  - Persisted model to '/tmp/tmppl2k3z4d/core'
Core model training completed.
Training NLU model...
2020-02-06 12:33:11 INFO     rasa.nlu.utils.spacy_utils  - Trying to load spacy model with name 'pt'
2020-02-06 12:33:12 INFO     rasa.nlu.components  - Added 'SpacyNLP' to component cache. Key 'SpacyNLP-pt'.
2020-02-06 12:33:12 INFO     rasa.nlu.training_data.training_data  - Training data stats:
        - intent examples: 2682 (226 distinct intents)
        - Found intents: ...
        - Number of response examples: 0 (0 distinct response)
        - entity examples: 39 (1 distinct entities)
        - found entities: ...

2020-02-06 12:33:12 INFO     rasa.nlu.model  - Starting to train component SpacyNLP
2020-02-06 12:33:15 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 12:33:15 INFO     rasa.nlu.model  - Starting to train component SpacyTokenizer
2020-02-06 12:33:15 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 12:33:15 INFO     rasa.nlu.model  - Starting to train component RegexFeaturizer
2020-02-06 12:33:15 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 12:33:15 INFO     rasa.nlu.model  - Starting to train component CRFEntityExtractor
2020-02-06 12:33:16 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 12:33:16 INFO     rasa.nlu.model  - Starting to train component CountVectorsFeaturizer
2020-02-06 12:33:17 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 12:33:17 INFO     rasa.nlu.model  - Starting to train component CountVectorsFeaturizer
2020-02-06 12:33:19 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 12:33:19 INFO     rasa.nlu.model  - Starting to train component tokenizer_spacy_lemma
2020-02-06 12:33:19 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 12:33:19 INFO     rasa.nlu.model  - Starting to train component EmbeddingIntentClassifier
Epochs: 100%|██████████████████████████████████| 220/220 [01:15<00:00,  2.91it/s, loss=1.326, acc=0.999]
2020-02-06 12:34:36 INFO     rasa.utils.train_utils  - Finished training embedding policy, train loss=1.326, train accuracy=0.999
2020-02-06 12:34:36 INFO     rasa.nlu.model  - Finished training component.
2020-02-06 12:34:37 INFO     rasa.nlu.model  - Successfully saved model into '/tmp/tmppl2k3z4d/nlu'
NLU model training completed.
```

**Comment by erohmensing** [comment]
@samscudder thanks for the detailed write up, it's really helpful. Looking into any changes that have been made to these components that could cause decreases in accuracy. A few follow up questions:

* What does your custom lemmatization do? Does it provide input to the intent classifier?
* Why do you have test input that is expected to fail? Why is it expected to fail, and is it part of the training data?

**Comment by samscudder** [comment]
I didn't write this , but the lemmatization script uses spacy training data to de-conjugate verbs (does that word exist?).

```
class CustomLemmatization(Tokenizer, Component):
    name = "tokenizer_spacy_lemma"
    provides = ["tokens"]
    requires = ["spacy_doc"]

    def train(self,
              training_data: TrainingData,
              config: RasaNLUModelConfig,
              **kwargs: Any)-> None:

        for example in training_data.training_examples:
            example.set("tokens", self.tokenize(example.get("spacy_doc")))

    def process(self, message: Message, **kwargs: Any)-> None:

        message.set("tokens", self.tokenize(message.get("spacy_doc")))

    def tokenize(self, doc: 'Doc')-> typing.List[Token]:
		
        return [Token(t.lemma_, t.idx) for t in doc]
```

Of the 102 phrases, I have 28 that are in the training data, 49 are close to  the training data, but can have a word mistyped or slightly rephrased and 25 that are grossly misspelled or don't make much sense grammatically..

The 25 expected to fail *are* failing (22 of them in Rasa 1.4.6 and all 25 in Rasa 1.7.0). A fail for me is a confidence < threshold or incorrect classification. So, where I said "9 phrases classified wrong", I mean some of the phrases expected to fail are being correctly classified, but have a confidence < threshold.

**Comment by erohmensing** [comment]
Right, but why do you expect them to fail? Shouldn't your model handle these cases by either incorporating those misspellings/grammar errors into your model, or handle them in e.g. an out-of-scope intent?

**Comment by samscudder** [comment]
Some of the grammar errors are expected to trigger the fallback, asking the user to rephrase his query. Phrases expected to fail are phrases a human operator wouldn't understand and would ask for clarification.

I'm sorry I can't post much about the dataset until our customer's site goes live. I've sent you the spreadsheet in the hope it demonstrates better.

**Comment by tabergma** [comment]
@samscudder Your `CustomLemmatization` is currently not used. It should be located before the featurizers. Also in Rasa 1.7 you just need to implement the `tokenize` method. `train` and `predict` is covered by the super class. However, as this is the case for both pipelines you used, it should not be the cause of error. 

Could you train Rasa 1.7 with the following pipeline:
```
pipeline:
- name: "SpacyNLP"
- name: "SpacyTokenizer"
- name: "RegexFeaturizer"
- name: "CRFEntityExtractor"
- name: "CountVectorsFeaturizer"
  strip_accents: "unicode"
- name: "CountVectorsFeaturizer"
  strip_accents: "unicode"
  analyzer: "char_wb"
  min_ngram: 1
  max_ngram: 4
- name: "components.lemmatization.lemma.CustomLemmatization"
  use_cls_token: false
- name: "EmbeddingIntentClassifier"
  epochs: 220
  random_seed: 2614
  ranking_length: 0
```
Just want to verify that is it not related to normalization. Please share the results of the training here. Thanks!

**Comment by samscudder** [comment]
Hi! Sorry for the delay.

With the above pipeline, training in 1.7 I get the following results:

The same 102 phrases, where 25 are purposely expected to fallback.
22 phrases classified wrong, 3 with confidence > 0.6
57 phrases were correctly classified, with a confidence > 0.6
Average overall confidence was 0.6362
Average confidence of the 57 phrases above 0.6 was 0,81

Overall it looks slightly worse than the previous 1.7.0

**Comment by tabergma** [comment]
Thanks for the numbers. We need to investigate this further. We will come back to you as soon as possible.

**Comment by samscudder** [comment]
Hi... any news on this?

**Comment by samscudder** [comment]
Rasa 1.7.4 was a big improvement in accuracy, with 75/102 correct above threshold (back to Rasa 1.4.6 levels), although the average accuracy of these 75 correctly classified is 8% lower than 1.4.6. 

Rasa 1.8 changed the pipeline, so I've been running some tests. I tested with 1 - 200 epochs in TEDPolicy and 100 - 300 epochs in DIETClassifier. The best results I've managed so far were with 200 TED epochs and 300 DIET. I haven't managed to get Tensorflow 2 to work with our GPU here, as it keeps aborting with cuda out of memory errors, and training is now terribly slow (6m -> 40m).

The results using TED and DIET weren't encouraging. With the same threshold (0.6) and the suggested pipeline for non-english, the number of correctly classified intents above threshold was only 34/102.

Using the same pipeline as 1.7.4, with EmbeddingIntentClassifier and Keras, training took 8 minutes (without GPU) but the average accuracy was around 0,15, and zero intents were above threshold.

**Comment by tabergma** [comment]
@samscudder Any chance you could share your data with us? We actually think that the data itself is the problem, but we cannot be sure without taking a closer look at them. As we changed quite a lot from 1.4 to 1.8 we would be happy to support you with finding a good setup for 1.8.1 that gives you the same/similar results as with 1.4, however, we don't have time to invest the performance of your dataset on older version. Sorry about that.

**Comment by samscudder** [comment]
After Rasa 1.9.6, the accuracy is at the same level or higher as Rasa 1.4.6
